### ISSUE TITLE
Construction of WHEREIS.HASH doesn't comprehend new lispusers sub-directory organization

### ISSUE THREAD

**Issue by nbriggs** [Interlisp/medley#741]
Lispusers packages that were moved to subdirectories do not appear to get indexed in the whereis.hash database.

**Comment by rmkaplan** [comment]
What about GATHER-INFO?  Is that also an issue?

**Comment by masinter** [comment]
https://github.com/Interlisp/medley/pull/720#pullrequestreview-902429411

**Comment by masinter** [comment]
my understanding of the semantics of emoji might be faulty. But it's interesting to think of whether Medley needs to support emoji if it's going to be current among linguistic tools.

The eyes I have chosen to mean "Hey, look at this and let's talk about it". I accept the risk that someone might take it to mean something else.

There are several possible solutions to the problem this issue raises.... For example:

* add a way of asking (find this file name on any level). 
* eliminate MEDLEYDIR and redo things using pseudohosts.
* whereishash could remember where it found the file in its index...

Anyway, in the short term setting it to 2 in make-whereis-hash would be fine.

GATHER-INFO is an unfinished project we should talk about; I lost momentum, too many open questions about how to make use of the info it has gathered.

**Comment by rmkaplan** [comment]
It seems that there are 2 separate issues:

1.  Make sure that the WHERE-IS scanning finds the file so that its functions can go into the database.  

2.  Make sure that the functions that want the definition of a function (LOAD, PF) can find it, given the information that is retrieved from the database even if the file is not yet known to the file package.

With the proper setting of search-path variables for both phases (DIRECTORIES, LISPUSERSDIRECTORIES), it ought to be the case that all the desired files are indexed and all the requested functions from those files can be located.

The new organizing principle of grouping the files of a multi-file lispusers (or library) package into a subdirectory may be screwing that up, because some files are systematically an extra level down.

XCL::WHERE-IS-FILES seems to be the guy that calls DIRECTORY to generate the list of files to be index.  So binding FILING.ENUMERATION.DEPTH to 2 might be enough to bring in files from their new locations.

FINDFNDEF is the function that tries to get the filemap for a file that the database tells it about.  It uses FINDFILE to find the file, which searches DIRECTORIES.  Would the files in the subdirectory be located if we changed FINDFILE to FINDFILE-WITH-EXTENSIONS?

In other words, would the short-term fix be just to set the indexing enumeration to 2 and change the retrieval to FINDFILE-FILE-WITH-EXTENSIONS?

Pseudohosts might be a better way of dealig with the relocatable MEDLEYDIR, I have been playing with that, and it cleans up a number of issues (you don’t need to have calls to MEDLEYDIR, the host name carries the site-relative location information).  But I don’t think that’s directly relevant to this issue, since we would still need to make sure that the indexing and the finding are synchronized.

**Comment by rmkaplan** [comment]
I changed the FINDFILE in FINDFNDEF to FINDFILE-WITH-EXTENSIONS, passing DIRECTORIES explicitly.

This doesn't work--everything gives a file-not-found error (without a break) if the EXTENSIONS list is NIL.  Essentially, this can't find source files, since they don't have an extension.

What would be the proper way of getting FINDFNDEF to know about the one-level-down files?

**Comment by masinter** [comment]
We have this hash file that maps from a symbol to what "managed file" it appeared in.
The only problem is the file list is "managed files only" and uses the root name only.

The thing that seems most general is to also remember where (relative to MEDLEYDIR) it found the file that it's noticing.

**Comment by rmkaplan** [comment]
Is there a cheap fix of just letting FINDFILE-WITH-EXTENSIONS interpret NIL or "" as the no-extension?  It seems that that would plug a hole in that interface, more general than the whereis problem

**Comment by masinter** [comment]
we should make a plan for adding file extensions to managed files.
in the meanwhile, do what you have to do.

**Comment by nbriggs** [comment]
It's not just `FINDFNDEF` that is an issue -- FILESLOAD doesn't work for files that are down a level either.  I'm not sure how FINDFILE-WITH-EXTENSIONS is helping -- it doesn't seem to enumerate subdirectories either (at least not in the sysout that I built on March 10th).

**Comment by rmkaplan** [comment]
I found a bug in the previous attempt at dealing with FOO-BAZ (using (IMINUS X 1) instead of (SUB1 x).  That made the original intent work for that new case (look in FOO>FOO-BAZ).  Separately, if NIL is in the list of extensions, the no-extension case would have worked before.  I fixed it so that EXTENSIONS defaults to (NIL).

What is an example of FILESLOAD failing, so I know what to look for?

**Comment by nbriggs** [comment]
`(FILESLOAD (FROM LISPUSERS) ACE)` doesn't find ACE/ACE or ACE/ACE.LCOM.  You can choose any file that is in a subdirectory of lispusers to show it.

**Comment by rmkaplan** [comment]
Two issues:

DOFILESLOAD was calling FINDFILE but not the expanded FINDFILE-WITH-EXTENSIONS.

But FINDFILE-WITH-EXTENSIONS had a condition that it wouldn’t do the extended search for FOO as FOO>FOO if a non-default explicit directory list (like LISPUSERS) was provided.

Larry, was there some reason (other than caution) for not allowing the descending search with an explicit directory list?

If I remove that condition, it does find ACE>ACE.  

----

The logic of this now is
     For each directory D
         For each extension E
              If D>file.E exists, return
              If D>file.E doesn’t exist and D>xfile>file.E exists, return  (xfile is the part of file before hyphen)

Would it be safer/better to do the descent in a separate loop, when explicit matching everywhere has failed?

     For each directory D
         For each extension E
              If D>file.E exists, return  (i.e., what would happen before this augmentation)
  If not found:
     For each directory D
         For each extension E
              If D>xfile>file.E exists, return

**Comment by nbriggs** [comment]
`SPELLFILE` may need to get fixed as well?  I don't think that having `FINDFILE` and `FINDFILE-WITH-EXTENSIONS` find different files is a good thing.   Any odd stuff that DOFILESLOAD does based on things like options `(SOURCE)` or `(COMPILED)` needs to work.

**Comment by rmkaplan** [comment]
Maybe FINDFILE should take an EXTENSIONS argument, and also have this new directory logic?

Very few callers of FINDFILE-WITH-EXTENSIONS, more with FINDFILE.  Maybe FINDFILE-WITH-EXTENSIONS should be removed (it’s not documented).

**Comment by nbriggs** [comment]
I see that FINDFILE's functionality is really just calling SPELLFILE and friends which have the complicated logic and account for oddities when you apply spelling correction for words to files with directories etc.  I haven't looked at all callers of FINDFILE-WITH-EXTENSIONS, but it was used in LAFITE in code that hasn't been modified since 1989, so I don't think it's a candidate for removal.   I think that everything I've noticed so far has been trying to find the compiled version of a file given a list of compiled file extensions.

**Comment by rmkaplan** [comment]
I’ve cleaned up the logic (but still doing the FOO>FOO in the inner directory/extension loop).

It works for ACE>ACE.  But there is also a confusion with that test case, because ACE itself does a (FILES (LOADCOMP) ACE) that is not in an ***@***.*** DONTCOPY.  So loading the LCOM also loadcomps the symbolics, and that leaves a whole bunch of stuff marked as changed.  I’m going to assume that that is a separate problem with ACE.

**Comment by masinter** [comment]
one of the problems is that this stuff needs to work early in the loadup; I had a lot of trouble with LOADUPDIRECTORIES and DIRECTORIES and LISPUSERSDIRECTORIES with loadups which were being run in different locations between my and Ron's configuration. Some of the problems might not be issues any more with the way we're doing loadups, but this code is fragile in more ways that is evident.   It was important to put MEDLEYDIR in the loadup in a way that the variables with full path names in them (like *DIRECTORIES) need to get updated before loading anything.

This bit of messiness one of the sources of my caution.

**Comment by nbriggs** [comment]
@rmkaplan -- I think in LOOPS I'm seeing similar issues: LOOPSKERNEL has a
```
(DECLARE%: EVAL@COMPILE DONTCOPY
     (FILES (LOADCOMP) ... LOOPSACTIVEVALUES ...)
)
```
and that also gets evaluated `@LOAD` so when the next `(DOFILESLOAD LOOPSACTIVEVALUES)` happens intending to load the source... it doesn't because it's already loaded enough as far as DOFILESLOAD is concerned.  Seems wrong -- should the LOADCOMPs be wrapped in a EVAL@COMPILE DONTEVAL@LOAD? What's the difference between EVAL@COMPILE and DOEVAL@COMPILE?

**Comment by rmkaplan** [comment]
Maybe you need  ***@***.*** ***@***.*** DONTCOPY ?

I think (hope?) that ***@***.*** and ***@***.*** are synonyms.

**Comment by rmkaplan** [comment]
Geez.

Maybe you need  ***@***.***` ***@***.***` DONTCOPY ?

I think (hope?) that ***@***.***` and ***@***.***` are synonyms.

**Comment by nbriggs** [comment]
Gah!  It must be the mail interface that does it, and it's ignoring the backquotes... the only way seems to be to say it directly on the web interface.   But, I guessed that you were suggesting EVAL(at)COMPILE and DONTEVAL(at)LOAD.   I tried that for the problem in LOOPSKERNEL and it fixed part of it but broke other things.   The dependencies (for both loading and compilation) are really NOT well specified.

**Comment by masinter** [comment]
I think it was partly my fault we never got the appendix to RFC1738 going forward.
Too much else to get right.
 https://datatracker.ietf.org/doc/html/rfc1738#:~:text=RFC%201738%20Uniform%20Resource%20Locators%20%28URL%29%20December%201994,in%20a%20scheme%2C%20the%20octet%20must%20be%20encoded.

**Comment by rmkaplan** [comment]
Trying again



Maybe you need  DONTEVALatLOAD EVALatCOMPILE DONTCOPY ?

I think (hope?) that DOEVALatCOMPILE and EVALatCOMPILE are synonyms.

**Comment by rmkaplan** [comment]
On further thought wrt ACE:  Does it ever make sense for a file to LOADCOMP itself?

**Comment by masinter** [comment]
This is another case in software archeology where there isn't an easy first-principle answer to "the right thing to do".
There's a bunch of false assumptions in the code about filename vs. FILENAME vs {DSK}<Users>Larry>FILENAME.;12.
so the right thing to do is to be clear about costs and benefits of each approach. 
File name lookup is expensive in directories with lots of files because there is some O(N^2) algorithms in directory enumeration from the Medley code.

**Comment by masinter** [comment]
I *think* the behavior that instigated this long thread has been fixed -- WHEREIS.HASH seems to find the files as long as we're careful to name them.

I'm marking this as 'closed' as fixed

**Comment by rmkaplan** [comment]
I don’t think speed is the primary consideration.  Much more costly not to find a file that you whose location you think you have properly specified.

FINDFILE has a quick test for INFILEP.  If that fails, I think it should then do the FINDFILE-WITH-EXTENSIONS (but only for the extension that the specified file has, if any), essentially just to enumerate the directories with the extra decoration of going down one level if that fails.  And then only do the whole SPELLFILE thing.
